File: Pastura/PasturaTests/Engine/TurnFailureGateTests.swift
import Foundation
import Testing

@testable import Pastura

@Suite(.timeLimit(.minutes(1)))
struct TurnFailureGateTests {
  private func skippedEvents(_ collector: EventCollector) -> [SimulationEvent] {
    collector.events.filter {
      if case .turnSkipped = $0 { return true }
      return false
    }
  }

  @Test func successReturnsValueAndEmitsNothing() async throws {
    let gate = TurnFailureGate()
    let collector = EventCollector()

    let value = try await gate.attempt(
      agent: "Alice", phaseType: .speakAll, emitter: collector.emit
    ) { "ok" }

    #expect(value == "ok")
    #expect(collector.events.isEmpty)
  }

  @Test func transientFailureSkipsTurnAndEmitsTurnSkipped() async throws {
    let gate = TurnFailureGate()
    let collector = EventCollector()

    let value: String? = try await gate.attempt(
      agent: "Alice", phaseType: .speakAll, emitter: collector.emit
    ) { throw SimulationError.retriesExhausted }

    #expect(value == nil)
    #expect(
      skippedEvents(collector) == [
        .turnSkipped(agent: "Alice", phaseType: .speakAll, cause: "retries exhausted")
      ])
  }

  @Test func generationFailureCarriesDescriptionAsCause() async throws {
    let gate = TurnFailureGate()
    let collector = EventCollector()

    let value: String? = try await gate.attempt(
      agent: "Bob", phaseType: .vote, emitter: collector.emit
    ) { throw SimulationError.llmGenerationFailed(description: "transient blip") }

    #expect(value == nil)
    #expect(
      skippedEvents(collector) == [
        .turnSkipped(agent: "Bob", phaseType: .vote, cause: "transient blip")
      ])
  }

  @Test func systemicErrorRethrowsTypedWithoutSkip() async throws {
    // ADR-021 D3: a deterministic engineering bug must abort in one throw,
    // not degrade turn-by-turn.
    let gate = TurnFailureGate()
    let collector = EventCollector()

    await #expect(throws: LLMError.invalidGrammar(description: "boom")) {
      _ = try await gate.attempt(
        agent: "Alice", phaseType: .speakAll, emitter: collector.emit
      ) { throw LLMError.invalidGrammar(description: "boom") }
    }
    #expect(collector.events.isEmpty)
  }

  @Test func cancellationRethrowsWithoutSkip() async throws {
    // ADR-021 D3 control-flow class: user cancellation must never be
    // converted into a skipped turn.
    let gate = TurnFailureGate()
    let collector = EventCollector()

    await #expect(throws: CancellationError.self) {
      _ = try await gate.attempt(
        agent: "Alice", phaseType: .speakAll, emitter: collector.emit
      ) { throw CancellationError() }
    }
    #expect(collector.events.isEmpty)
  }

  @Test func thirdConsecutiveFailureTripsBreaker() async throws {
    // ADR-021 D4: skips 1 and 2 emit .turnSkipped; the 3rd consecutive
    // failure throws turnFailureLimitReached INSTEAD of a 3rd skip. The
    // counter is shared across phases — the failures span speak_all → vote.
    let gate = TurnFailureGate()
    let collector = EventCollector()

    for phase in [PhaseType.speakAll, .speakAll] {
      let value: String? = try await gate.attempt(
        agent: "Alice", phaseType: phase, emitter: collector.emit
      ) { throw SimulationError.retriesExhausted }
      #expect(value == nil)
    }

    await #expect(throws: SimulationError.turnFailureLimitReached(consecutiveCount: 3)) {
      _ = try await gate.attempt(
        agent: "Bob", phaseType: .vote, emitter: collector.emit
      ) { throw SimulationError.retriesExhausted }
    }
    // Only the first two failures produced skip events.
    #expect(skippedEvents(collector).count == 2)
  }

  @Test func successResetsConsecutiveCounter() async throws {
    // ADR-021 D4: F F S F F must NOT trip the 3-consecutive breaker.
    let gate = TurnFailureGate()
    let collector = EventCollector()

    func fail() async throws -> String? {
      try await gate.attempt(
        agent: "Alice", phaseType: .speakAll, emitter: collector.emit
      ) { throw SimulationError.retriesExhausted }
    }

    _ = try await fail()
    _ = try await fail()
    let ok = try await gate.attempt(
      agent: "Alice", phaseType: .speakAll, emitter: collector.emit
    ) { "recovered" }
    #expect(ok == "recovered")
    _ = try await fail()
    _ = try await fail()

    #expect(skippedEvents(collector).count == 4)
  }
}

File: shared/engine/src/commonTest/kotlin/com/pastura/engine/TurnFailureGateTests.kt
package com.pastura.engine

import com.pastura.models.PhaseType
import com.pastura.models.SimulationError
import com.pastura.models.SimulationEvent
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlin.test.assertFailsWith

/**
 * commonTest parity spec for [TurnFailureGate], ported from
 * `Pastura/PasturaTests/Engine/TurnFailureGateTests.swift` (7 Swift tests) plus
 * one new wrapped-wrong-case probe.
 */
class TurnFailureGateTests {

    // A type DISJOINT from the cancellation hierarchy: Kotlin
    // `CancellationException : IllegalStateException`, so a systemic-rethrow probe
    // must not use IllegalStateException or it would read as cancellation.
    private class ProbeError : RuntimeException("boom")

    private fun MutableList<SimulationEvent>.skipped(): List<SimulationEvent.TurnSkipped> =
        filterIsInstance<SimulationEvent.TurnSkipped>()

    @Test
    fun successReturnsValueAndEmitsNothing() = runTest {
        val gate = TurnFailureGate()
        val events = mutableListOf<SimulationEvent>()

        val value = gate.attempt(
            agent = "Alice", phaseType = PhaseType.SPEAK_ALL, emitter = { events.add(it) },
        ) { "ok" }

        assertEquals("ok", value)
        assertTrue(events.isEmpty())
    }

    @Test
    fun transientFailureSkipsTurnAndEmitsTurnSkipped() = runTest {
        val gate = TurnFailureGate()
        val events = mutableListOf<SimulationEvent>()

        val value: String? = gate.attempt(
            agent = "Alice", phaseType = PhaseType.SPEAK_ALL, emitter = { events.add(it) },
        ) { throw SimulationException(SimulationError.RetriesExhausted) }

        assertNull(value)
        assertEquals(
            listOf(
                SimulationEvent.TurnSkipped(
                    agent = "Alice", phaseType = PhaseType.SPEAK_ALL, cause = "retries exhausted",
                ),
            ),
            events.skipped(),
        )
    }

    @Test
    fun generationFailureCarriesDescriptionAsCause() = runTest {
        val gate = TurnFailureGate()
        val events = mutableListOf<SimulationEvent>()

        val value: String? = gate.attempt(
            agent = "Bob", phaseType = PhaseType.VOTE, emitter = { events.add(it) },
        ) { throw SimulationException(SimulationError.LlmGenerationFailed("transient blip")) }

        assertNull(value)
        assertEquals(
            listOf(
                SimulationEvent.TurnSkipped(
                    agent = "Bob", phaseType = PhaseType.VOTE, cause = "transient blip",
                ),
            ),
            events.skipped(),
        )
    }

    @Test
    fun systemicErrorRethrowsTypedWithoutSkip() = runTest {
        // ADR-021 D3: a deterministic engineering bug must abort in one throw,
        // not degrade turn-by-turn. ProbeError stands in for Swift's
        // LLMError.invalidGrammar (not ported to Kotlin).
        val gate = TurnFailureGate()
        val events = mutableListOf<SimulationEvent>()

        assertFailsWith<ProbeError> {
            gate.attempt(
                agent = "Alice", phaseType = PhaseType.SPEAK_ALL, emitter = { events.add(it) },
            ) { throw ProbeError() }
        }
        assertTrue(events.isEmpty())
    }

    @Test
    fun cancellationRethrowsWithoutSkip() = runTest {
        // ADR-021 D3 control-flow class: user cancellation must never be
        // converted into a skipped turn.
        val gate = TurnFailureGate()
        val events = mutableListOf<SimulationEvent>()

        assertFailsWith<CancellationException> {
            gate.attempt(
                agent = "Alice", phaseType = PhaseType.SPEAK_ALL, emitter = { events.add(it) },
            ) { throw CancellationException("cancel") }
        }
        assertTrue(events.isEmpty())
    }

    @Test
    fun thirdConsecutiveFailureTripsBreaker() = runTest {
        // ADR-021 D4: skips 1 and 2 emit TurnSkipped; the 3rd consecutive failure
        // throws turnFailureLimitReached INSTEAD of a 3rd skip. The counter is
        // shared across phases — the failures span speak_all → vote.
        val gate = TurnFailureGate()
        val events = mutableListOf<SimulationEvent>()

        for (phase in listOf(PhaseType.SPEAK_ALL, PhaseType.SPEAK_ALL)) {
            val value: String? = gate.attempt(
                agent = "Alice", phaseType = phase, emitter = { events.add(it) },
            ) { throw SimulationException(SimulationError.RetriesExhausted) }
            assertNull(value)
        }

        val thrown = assertFailsWith<SimulationException> {
            gate.attempt(
                agent = "Bob", phaseType = PhaseType.VOTE, emitter = { events.add(it) },
            ) { throw SimulationException(SimulationError.RetriesExhausted) }
        }
        val error = thrown.error
        assertIs<SimulationError.TurnFailureLimitReached>(error)
        assertEquals(3, error.consecutiveCount)
        // Only the first two failures produced skip events.
        assertEquals(2, events.skipped().size)
    }

    @Test
    fun successResetsConsecutiveCounter() = runTest {
        // ADR-021 D4: F F S F F must NOT trip the 3-consecutive breaker.
        val gate = TurnFailureGate()
        val events = mutableListOf<SimulationEvent>()

        suspend fun fail(): String? = gate.attempt(
            agent = "Alice", phaseType = PhaseType.SPEAK_ALL, emitter = { events.add(it) },
        ) { throw SimulationException(SimulationError.RetriesExhausted) }

        fail()
        fail()
        val ok = gate.attempt(
            agent = "Alice", phaseType = PhaseType.SPEAK_ALL, emitter = { events.add(it) },
        ) { "recovered" }
        assertEquals("recovered", ok)
        fail()
        fail()

        assertEquals(4, events.skipped().size)
    }

    @Test
    fun wrappedNonDegradableCaseRethrowsBeforeCounterIncrements() = runTest {
        // NEW (fills a gap the Swift set misses): a WRAPPED but non-degradable
        // SimulationError (ModelNotLoaded) must rethrow before the counter
        // increments — proving the branch a too-broad `catch(SimulationException)`
        // would silently degrade. No skip event may be emitted.
        val gate = TurnFailureGate()
        val events = mutableListOf<SimulationEvent>()

        val thrown = assertFailsWith<SimulationException> {
            gate.attempt(
                agent = "Alice", phaseType = PhaseType.SPEAK_ALL, emitter = { events.add(it) },
            ) { throw SimulationException(SimulationError.ModelNotLoaded) }
        }
        assertIs<SimulationError.ModelNotLoaded>(thrown.error)
        assertTrue(events.isEmpty())
    }
}
